File: js/stats.js
/* jshint esversion: 6, devel: true, browser: true */

const Stats = {
   get lifetimeBlocksCollected() {
      return this.lifetimeBlocksProducedByFactories + Data.lifetimeBlocksByColor.red + Data.lifetimeBlocksByColor.orange + Data.lifetimeBlocksByColor.green + Data.lifetimeBlocksByColor.blue;
   },
   get lifetimeBlocksProducedByFactories() {
      let total = 0;
      for (let prop in factories) {
         total += factories[prop].totalBlocksProduced;
      }
      return total;
   },
   get favoriteColor() {
      let max = Math.max(Data.lifetimeBlocksByColor.red, Math.max(Data.lifetimeBlocksByColor.orange, Math.max(Data.lifetimeBlocksByColor.green, Math.max(Data.lifetimeBlocksByColor.blue, Math.max(Data.lifetimeBlocksByColor.toxic, Data.lifetimeBlocksByColor.golden)))));
      switch (max) {
         case Data.lifetimeBlocksByColor.red:
            return "Red";
         case Data.lifetimeBlocksByColor.orange:
            return "Orange";
         case Data.lifetimeBlocksByColor.green:
            return "Green";
         case Data.lifetimeBlocksByColor.blue:
            return "Blue";
         case Data.lifetimeBlocksByColor.toxic:
            return "Toxic blocks";
         case Data.lifetimeBlocksByColor.golden:
            return "Golden blocks";
      }
   }
};

/** Returns the y-coordinate where the next statistic can be rendered */
const renderBigStat = (statName, statValue, ctx2d, textY, maxWidth) => {
   ctx2d.fillStyle = 'black';
   ctx2d.textBaseline = 'top';
   ctx2d.textAlign = 'center';

   let bigFontSize = UI_SANS_TEXT_HEIGHT * 1.25;
   let bigSansSerif = toBrowserH(bigFontSize) + 'px sans-serif';
   ctx2d.font = bigSansSerif;
   ctx2d.fillText(statName, toBrowserX(VISIBLE_WIDTH / 2), toBrowserY(textY));

   textY += bigFontSize;
   ctx2d.font = toBrowserH(75) + 'px Digital-7';
   ctx2d.fillText(statValue, toBrowserX(VISIBLE_WIDTH / 2), toBrowserY(textY), maxWidth);

   return textY + 75 + UI_PADDING;
};

/** Returns the y-coordinate where the next statistic can be rendered */
const renderSmallStat = (statName, statValue, ctx2d, textY) => {
   ctx2d.fillStyle = "black";
   ctx2d.textBaseline = "alphabetic";

   ctx2d.textAlign = "left";
   ctx2d.fillText(statName, toBrowserX(UI_PADDING), toBrowserY(textY));

   ctx2d.textAlign = "right";
   ctx2d.fillText(statValue, toBrowserX(VISIBLE_WIDTH - UI_PADDING), toBrowserY(textY));

   ctx2d.beginPath();
   ctx2d.setLineDash([1, 3]);
   ctx2d.moveTo(toBrowserX(UI_PADDING + 5) + ctx2d.measureText(statName).width, toBrowserY(textY));
   ctx2d.lineTo(toBrowserX(VISIBLE_WIDTH - UI_PADDING - 5) - ctx2d.measureText(statValue).width, toBrowserY(textY));
   ctx2d.stroke();
   return textY + UI_SANS_TEXT_HEIGHT * 1.25;
};

const renderStats = (ctx2d, yOffset) => {
   let maxWidth = toBrowserW(VISIBLE_WIDTH - UI_PADDING * 2);
   let textY = getStatusBarHeight() + UI_PADDING + yOffset;

   textY = renderBigStat('Lifetime blocks produced or collected', Math.floor(Stats.lifetimeBlocksCollected).toLocaleString(), ctx2d, textY, maxWidth);
   textY = renderBigStat('Lifetime pollutants produced', Math.floor(Data.lifetimePollution).toLocaleString(), ctx2d, textY, maxWidth);

   /* Render small stats */
   textY += UI_PADDING;

   let total = 0;
   for (let prop in Data.lifetimeBlocksByColor)
      total += Data.lifetimeBlocksByColor[prop];

   ctx2d.font = getSansFont();
   ctx2d.strokeStyle = "black";
   ctx2d.lineWidth = 1;

   // Block stats
   textY = renderSmallStat("Lifetime blocks collected by hand", total.toLocaleString(), ctx2d, textY);
   textY = renderSmallStat("Red blocks", Data.lifetimeBlocksByColor.red.toLocaleString(), ctx2d, textY);
   textY = renderSmallStat("Orange blocks", Data.lifetimeBlocksByColor.orange.toLocaleString(), ctx2d, textY);
   textY = renderSmallStat("Green blocks", Data.lifetimeBlocksByColor.green.toLocaleString(), ctx2d, textY);
   textY = renderSmallStat("Blue blocks", Data.lifetimeBlocksByColor.blue.toLocaleString(), ctx2d, textY);

   // Adds a gap
   textY += UI_SANS_TEXT_HEIGHT * 1.25;

   // Click stats
   textY = renderSmallStat("Successful block clicks", Data.lifetimeClicks.successful.toLocaleString(), ctx2d, textY);
   textY = renderSmallStat("Failed block clicks", Data.lifetimeClicks.failed.toLocaleString(), ctx2d, textY);

   // Adds a gap
   textY += UI_SANS_TEXT_HEIGHT * 1.25;

   // Factory stats
   let totalFactories = 0;
   for (let prop in factories)
      totalFactories += factories[prop].amountOwned;
   textY = renderSmallStat("Total factories owned", totalFactories, ctx2d, textY);
   textY = renderSmallStat("Lifetime blocks produced by factories", Math.floor(Stats.lifetimeBlocksProducedByFactories).toLocaleString(), ctx2d, textY);

};

File: src/stats.rs
use crate::factories::FACTORIES;
use crate::data::Data;
use std::fmt::Display;

use crate::ui;

struct Stats {

}

impl Stats {
    fn lifetime_blocks_collected(&self, data: Data) -> u128 {
        self.lifetime_blocks_produced_by_factories + data.lifetime_blocks_by_color.red + data.lifetime_blocks_by_color.orange + data.lifetime_blocks_by_color.green + data.lifetime_blocks_by_color.blue;
    }

    fn lifetime_blocks_produced_by_factories() {
        let total = 0;
        for f in FACTORIES {
            total += factories[prop].total_blocks_produced;
        }
        total
    }

    fn favorite_color(data: &Data) {
        let max = data.lifetime_blocks_by_color.red.max(data.lifetime_blocks_by_color.orange.max(data.lifetime_blocks_by_color.green).max(data.lifetime_blocks_by_color.blue.max(data.lifetime_blocks_by_color.toxic.max(data.lifetime_blocks_by_color.golden))));
            if max == data.lifetime_blocks_by_color.red { "Red" }
            else if max == data.lifetime_blocks_by_color.orange { "Orange" }
            else if max == data.lifetime_blocks_by_color.green { "Green" }
            else if max == data.lifetime_blocks_by_color.blue { "Blue" }
            else if max == data.lifetime_blocks_by_color.toxic { "Toxic blocks" }
            else if max == data.lifetime_blocks_by_color.golden { "Golden blocks" }
            else { "Unknown" }
    }
}

/// Returns the y-coordinate where the next statistic can be rendered
fn render_big_stat<CanvasContext>(stat_name: &str, stat_value: impl Display, ctx_2d: CanvasContext, text_y: f32, max_width: f32) {
    ctx_2d.fill_style = "black";
    ctx_2d.text_baseline = "top";
    ctx_2d.text_align = "center";

    let big_font_size = ui::SANS_TEXT_HEIGHT * 1.25;
    let big_sans_serif = format!("{}px sans-serif", to_browser_h(big_font_size));
    ctx_2d.font = bigSansSerif;
    ctx_2d.fill_text(stat_name, utils::to_browser_x(VISIBLE_WIDTH / 2), to_browser_y(text_y));

    text_y += big_font_size;
    ctx_2d.font = format!("{}px Digital-7", utils::to_browser_h(75));
    ctx_2d.fill_text(statValue, to_browser_x(VISIBLE_WIDTH / 2), to_browser_y(text_y), max_width);

    text_y + 75 + ui::PADDING
}

/** s the y-coordinate where the next statistic can be rendered */
fn render_small_stat<CanvasContext>(stat_name: &str, stat_value: impl Display, ctx_2d: CanvasContext, text_y: f32) {
    ctx_2d.fill_style = "black";
    ctx_2d.text_baseline = "alphabetic";

    ctx_2d.text_align = "left";
    ctx_2d.fill_text(stat_name, to_browser_x(UI_PADDING), to_browser_y(text_y));

    ctx_2d.text_align = "right";
    ctx_2d.fill_text(stat_value, to_browser_x(VISIBLE_WIDTH - UI_PADDING), to_browser_y(text_y));

    ctx_2d.begin_path();
    ctx_2d.set_line_dash([1, 3]);
    ctx_2d.move_to(to_browser_x(ui::PADDING + 5) + ctx_2d.measure_text(statName).width, to_browser_y(text_y));
    ctx_2d.line_to(to_browser_x(VISIBLE_WIDTH - UI_PADDING - 5) - ctx_2d.measure_text(stat_value).width, to_browser_y(text_y));
    ctx_2d.stroke();
    text_y + UI_SANS_TEXT_HEIGHT * 1.25
}

fn render_stats<CanvasContext>(ctx_2d: CanvasContext, y_offset: f32) {
    let max_width = to_browser_w(VISIBLE_WIDTH - UI_PADDING * 2);
    let text_y = get_status_bar_height() + UI_PADDING + y_offset;

    text_y = render_big_stat("All blocks produced or collected", Math.floor(Stats.lifetime_blocks_collected).to_locale_string(), ctx_2d, text_y, max_width);
    // text_y = render_big_stat("Lifetime pollutants produced", Math.floor(Data.lifetime_pollution).to_locale_string(), ctx_2d, text_y, max_width);

    // render small stats
    text_y += UI_PADDING;

    let total = data.get_total_blocks();

    ctx_2d.font = getSansFont();
    ctx_2d.strokeStyle = "black";
    ctx_2d.lineWidth = 1;

    // Block stats
    text_y = render_small_stat("All blocks collected by hand", total.to_locale_string(), ctx_2d, text_y);
    text_y = render_small_stat("Red blocks", Data.lifetime_blocks_by_color.red.to_locale_string(), ctx_2d, text_y);
    text_y = render_small_stat("Orange blocks", Data.lifetime_blocks_by_color.orange.to_locale_string(), ctx_2d, text_y);
    text_y = render_small_stat("Green blocks", Data.lifetime_blocks_by_color.green.to_locale_string(), ctx_2d, text_y);
    text_y = render_small_stat("Blue blocks", Data.lifetime_blocks_by_color.blue.to_locale_string(), ctx_2d, text_y);

    // Adds a gap
    text_y += UI_SANS_TEXT_HEIGHT * 1.25;

    // Click stats
    text_y = render_small_stat("Successful block clicks", Data.lifetime_clicks.successful.to_locale_string(), ctx_2d, text_y);
    text_y = render_small_stat("Failed block clicks", Data.lifetime_clicks.failed.to_locale_string(), ctx_2d, text_y);

    // Adds a gap
    text_y += UI_SANS_TEXT_HEIGHT * 1.25;

    // Factory stats
    let total_factories = 0;
    for f in factories {
        total_factories += factories[prop].amount_owned;
    }
    text_y = render_small_stat("Total factories owned", total_factories, ctx_2d, text_y);
    text_y = render_small_stat("All blocks produced by factories", (Stats.lifetime_blocks_produced_by_factories).floor().to_locale_string(), ctx_2d, text_y);

}
